text/0000-native-class-constructor-update.md

```diff
@@ -0,0 +1,220 @@
+
+```js
+class Foo extends EmberObject {
+  // Does not work, for the reasons described above
+  bar = 'baz';
+
+  // Does not cover all cases. If we did `Foo.create({ bar: false })` it would
+  // still assign the default.
+  bar = this.bar || 'baz';
+
+  // This works, but is very verbose and not ideal
+  bar = this.hasOwnProperty('bar') ? this.bar : 'baz';
+
+  // This is one of the community accepted solutions, but it requires lodash
+  bar = _.defaultTo(this.bar, 'baz');
+
+  // This is another community accepted solution, but it requires
+  // @ember-decorators/argument, which is a separate library
+  @argument foo = 'bar';
+}
+```
+
+None of these is ideal. Instead, we can change the behavior of the constructor
+and the `create` method to circumvent this issue.
+
+This change _would_ be a breaking change to the behavior of native classes
+today, and a change from the previous class RFC. This will impact early adopters
+and should be made with that in mind. However, because native classes never
+officially shipped as part of Ember's public API (an announcement was not made,
+docs have not been written, etc), this RFC proposes that the change would _not_
+be considered a breaking change _for the purposes of semver_. This would allow
+us to ship the change during the Ember v3 release cycle, and prevent more code
+from being built on top of the previous behavior.
+
+## Detailed design
+
+One very important design constraint to making this change is that we _cannot_
+break the behavior of EmberObject when used _without_ native classes. To do
+this, we will leverage the fact that the static `create` method is the only
+public way to create an instance of EmberObject.
+
+Currently, the behavior of EmberObject is the following (simplified):
+
+```js
+class EmberObject {
+  constructor(props) {
+    // ..class setup things
+
+    Object.assign(this, props);
+    this.init();
+  }
+
+  static create(props) {
+    let instance = new this(props);
+
+    return instance;
+  }
+}
+```
+
+We can change it to the following (simplified):
+
+```js
+class EmberObject {
+  constructor(props) {
+    // ..class setup things
+  }
+
+  static create(props) {
+    let instance = new this(props);
+
+    Object.assign(this, props);
+    this.init();
+
+    return instance;
+  }
+}
+```
+
+This would assign the properties _after_ all of the class fields for any
+subclasses have been assigned. Revisiting our previous example, the following
+two class declarations would effectively be equivalent:
+
+```js
+const Foo = EmberObject.extend({
+  bar: 'baz'
+});
+
+class Foo extends EmberObject {
+  bar = 'baz';
+}
+```
+
+Much easier to codemod! There are other subtle differences between native class
+fields and EmberObject properties, such as the fact that class fields are
+assigned each time a class is initialized, but these are easier to work around.
+
+### Injections and the `init` hook
+
+One side effect of this change is that injections will not be available on the
```

Comment:
It's worth noting that we've done an unofficial survey of some of the largest apps *known* to be using ES6 classes in production via TypeScript and have found something like 2 instances of the constructor using injected services *total* across them. I don't think we have enough data to say that in the RFC, but I wanted to bring it up here for the record in the RFC discussion.